### ISSUE TITLE
inconsistent format of results of EntsoePandasClient.query_generation()

### ISSUE THREAD

**Issue by mkaut** [EnergieID/entsoe-py#328]
I am testing the `EntsoePandasClient` and getting inconsistent formatting of results of `query_generation()`, in several ways.
In all cases, I am asking for data from 2019 to 2023, i.e., I am calling `client.query_generation(country_code, start=pd.Timestamp('20190101', tz='Europe/Brussels'), end=pd.Timestamp('20240101', tz='Europe/Brussels'))`.

For Germany (`country_code = 'DE_LU'`), the result has a multi-indexed columns:
```
                                    Biomass Fossil Brown coal/Lignite Fossil Coal-derived gas        Fossil Gas                     ...              Solar             Waste     Wind Offshore      Wind Onshore
                          Actual Aggregated         Actual Aggregated       Actual Aggregated Actual Aggregated Actual Consumption  ... Actual Consumption Actual Aggregated Actual Aggregated Actual Aggregated Actual Consumption
2019-01-01 00:00:00+01:00            4812.0                    6932.0                   273.0            3410.0                1.0  ...                NaN             783.0            3177.0           19366.0                NaN
2019-01-01 00:15:00+01:00            4828.0                    6351.0                   481.0            3295.0                1.0  ...                NaN             772.0            3174.0           20132.0                NaN
2019-01-01 00:30:00+01:00            4834.0                    6221.0                   481.0            3228.0                1.0  ...                NaN             779.0            3167.0           20863.0                NaN
```

The same query for Denmark's DK-1 zone (`country_code='DK_1'`) returns single-indexed columns:
```
                           Biomass  (Biomass, Actual Aggregated)  (Biomass, Actual Consumption)  ...  (Wind Offshore, Actual Consumption)  (Wind Onshore, Actual Aggregated)  (Wind Onshore, Actual Consumption)
2019-01-01 00:00:00+01:00      NaN                          79.0                            NaN  ...                                  NaN                             2330.0                                 NaN
2019-01-01 01:00:00+01:00      NaN                          62.0                            NaN  ...                                  NaN                             2427.0                                 NaN
2019-01-01 03:00:00+01:00      NaN                          62.0                            NaN  ...                                  NaN                             2290.0                                 NaN
2019-01-01 04:00:00+01:00      NaN                          58.0                            NaN  ...                                  NaN                             2229.0                                 NaN
```
Note that the column names are tuples, but the `.columns` is still an `Index`, not `MultiIndex` like for Germany.

What's even worse, the column assignment changes when I use the `psr_type` argument in the call. To illustrate this, consider all columns for offshore wind from the previous dataframe:
```
                           Wind Offshore  (Wind Offshore, Actual Aggregated)  (Wind Offshore, Actual Consumption)
2019-01-01 00:00:00+01:00            NaN                               638.0                                  NaN
2019-01-01 01:00:00+01:00            NaN                               686.0                                  NaN
2019-01-01 02:00:00+01:00            NaN                               296.0                                  NaN
2019-01-01 03:00:00+01:00            NaN                               289.0                                  NaN
2019-01-01 04:00:00+01:00            NaN                               283.0                                  NaN
...                                  ...                                 ...                                  ...
2023-12-31 19:00:00+01:00         1129.0                                 NaN                                  NaN
2023-12-31 20:00:00+01:00         1093.0                                 NaN                                  NaN
2023-12-31 21:00:00+01:00         1165.0                                 NaN                                  NaN
2023-12-31 22:00:00+01:00         1191.0                                 NaN                                  NaN
2023-12-31 23:00:00+01:00         1163.0                                 NaN                                  NaN
```
There, we can see that the values actually switch columns somewhere during the period.
**EDIT:** It turns out the data switch column several times: they are in `(Wind Offshore, Actual Aggregated)` in 2019 and 2021 and in `Wind Offshore` in 2020, 2022, and 2023.
Also note the inconsistency in naming, with the first column having name as string, while the other two as a tuple.

On the other hand, asking only for offshore wind with `psr_type='B18'` returns
```
                           (Wind Offshore, Actual Aggregated)  (Wind Offshore, Actual Consumption)  Wind Offshore
2019-01-01 00:00:00+01:00                                 NaN                                  NaN          638.0
2019-01-01 01:00:00+01:00                                 NaN                                  NaN          686.0
2019-01-01 02:00:00+01:00                                 NaN                                  NaN          296.0
2019-01-01 03:00:00+01:00                                 NaN                                  NaN          289.0
2019-01-01 04:00:00+01:00                                 NaN                                  NaN          283.0
...                                                       ...                                  ...            ...
2023-12-31 19:00:00+01:00                                 NaN                                  NaN         1129.0
2023-12-31 20:00:00+01:00                                 NaN                                  NaN         1093.0
2023-12-31 21:00:00+01:00                                 NaN                                  NaN         1165.0
2023-12-31 22:00:00+01:00                                 NaN                                  NaN         1191.0
2023-12-31 23:00:00+01:00                                 NaN                                  NaN         1163.0
```
i.e., the values are in the `Wind Offshore column` in all years.
**EDIT:** The values turned out to be in column `(Wind Offshore, Actual Aggregated)` in 2021.

In other words, values one gets with the `psr_type` argument are **not** a subset of values without, as I would expect.

**Comment by fboerman** [comment]
Please provide some examples with code snippets and clear explanation

-------- Original Message --------

**Comment by fboerman** [comment]
@mkaut  hi I need a direct code snippit which I can paste into my console to execute, thanks!

**Comment by BerriJ** [comment]
Here is a reprex for this issue:

```
# %% Setup
from entsoe import EntsoePandasClient
import pandas as pd

client = EntsoePandasClient(api_key="072d42a6-043a-48cf-b909-b571f0d9ed1c")

# %% Get data
DE_LU = client.query_generation(
    "DE_LU",
    start=pd.Timestamp("20250101", tz="Europe/Brussels"),
    end=pd.Timestamp("20250103", tz="Europe/Brussels"),
)
DK_1 = client.query_generation(
    "DK_1",
    start=pd.Timestamp("20250101", tz="Europe/Brussels"),
    end=pd.Timestamp("20250103", tz="Europe/Brussels"),
)

# Inspect index
print(DE_LU.columns)
print(DK_1.columns)
```

Indeed, `DE_LU.columns` is a MultiIndex while `DK_1.columns` is not.

**Comment by BerriJ** [comment]
Thank you for the clarification (I already suspected the API). This is unfortunate as consistency matters *a lot* when dealing with these kinds of data.

My workaround is this:

```python
if not isinstance(df.columns, pd.MultiIndex):
    result[i].columns = pd.MultiIndex.from_tuples(
        [(col, "Actual Aggregated") for col in df.columns],
        names=[df.columns.name or None, None],
    )
```

My interpretation is: If there are generation types that not only generate but *also* consume, then entsoe adds a multi-index that differentiates "Actual Aggregated" (Generation) from "Actual Consumption" (Consumption). So when we do not have a multi-index, we can deduce that all columns must be Generation. Therefore, I'm adding "Actual Aggregated" to all of them so I have consistency across all tables.

The question is whether Entsoe-Py should handle this or not. The cleanest solution would be to fix it on the entsoe side.

**Comment by fboerman** [comment]
well I have no influence on what entsoe does, so can only fix it in the package. your solution might work, can you create a pull request with a clear code to test the change? Thanks!

**Comment by morrisonmaxw** [comment]
Hi team, thought I would share my own workaround here in-case it is helpful to anyone. This function pivots and flattens the dataframe into a consistent structure regardless of the original format. Suggest any fix be backwards compatible to not break people's workarounds. 
-Max

`    # Function to process and pivot data
def process_generation_dataframe(Entsoe_client_request: pd.DataFrame, country_code: str):
    # Reset index to move timestamp to a column
    Entsoe_client_request = Entsoe_client_request.reset_index()
    logging.debug(f"Columns after reset_index:\n{Entsoe_client_request.columns}")

    if isinstance(Entsoe_client_request.columns, pd.MultiIndex):
        df = Entsoe_client_request
        # Rename the index column to 'Timestamp_UTC' and handle the second level
        df.columns = pd.MultiIndex.from_tuples(
            [('Timestamp_UTC', '') if col[0] == 'index' else (col[0], col[1].replace(' ', '_')) for col in Entsoe_client_request.columns]
        )
        logging.debug(f"Columns after renaming index:\n{df.columns}")
        # Set Timestamp_UTC as index
        
        df.set_index(('Timestamp_UTC', ''), inplace=True)

        # Flatten the MultiIndex DataFrame
        df_flat = df.stack(level=0, future_stack=True).reset_index()

        # Rename columns
        df_flat.columns = ['Timestamp_UTC', 'Production_Type', 'Actual_Aggregated_Generation_MW', 'Actual_Consumption_Generation_MW']

        # Display the flattened DataFrame
        logging.debug(f"Dataframe columns after flattening:\n{df_flat.columns}")
        logging.debug(f"Sample flattened dataframe:\n{df_flat.head()}")

        processed_dataframe = df_flat

    else:
        # Rename the index column to 'Timestamp_UTC' for single index columns
        df = Entsoe_client_request
        df.rename(columns={'index': 'Timestamp_UTC'}, inplace=True)
        logging.debug(f"Columns after renaming (SingleIndex):\n{df.columns}")
        logging.debug(f"Sample simple dataframe after renaming:\n{df.head()}")
        
        df_pivot = df.melt(id_vars=['Timestamp_UTC'], var_name='Production_Type', value_name='Actual_Aggregated_Generation_MW')
        df_pivot.insert(3, 'Actual_Consumption_Generation_MW', pd.NA)
        processed_dataframe=df_pivot
    # Insert Country_Code column for context
    processed_dataframe.insert(1, 'Country_Code', country_code)

    return processed_dataframe`
